### PR TITLE
Demonstrate the correct way to send a welcome message in 1:1

### PR DESCRIPTION
**Pull request by aosolis**
* Detect condition where bot is added to 1:1 or team
* Check for previous welcome message before sending welcome to 1:1 chat
* Implement /resetbotchat to simulate a new conversation. Teams does not support re-creating a 1:1 chat, so the recommended approach is to simulate it.

### CODE REVIEW

File: template-bot-master-csharp/src/controllers/MessagesController.cs [hunk 1]
```diff
@@ -136,12 +140,60 @@ private Activity HandleSystemMessage(Activity message)
...
-                //uncomment the below line to handle cnversation update messages
-                //TeamEventBase eventData = message.GetConversationUpdateData();
+                // This shows how to send a welcome message in response to a conversationUpdate event
+
+                // We're only interested in member added events
+                if (message.MembersAdded?.Count > 0)
+                {
+                    // Determine if the bot was added to the team/conversation
+                    var botId = message.Recipient.Id;
+                    var botWasAdded = message.MembersAdded.Any(member => member.Id == message.Recipient.Id);
```
Comment (msftdherron) [inline template-bot-master-csharp/src/controllers/MessagesController.cs]: use your new botId variable?

File: template-bot-master-csharp/src/controllers/MessagesController.cs [hunk 2]
```diff
@@ -136,12 +140,60 @@ private Activity HandleSystemMessage(Activity message)
...
+                            {
+                                await connectorClient.Conversations.ReplyToActivityWithRetriesAsync(welcomeMessage, cancellationToken);
+
+                                // Remember that we sent the welcome message already
+                                botData.SetProperty("IsFreSent", true);
+                                await botDataStore.SaveAsync(address, BotStoreType.BotUserData, botData, cancellationToken);
+                            }
+                            else
+                            {
+                                // First-run message has already been sent, so skip sending it again
```
Comment (msftdherron) [inline template-bot-master-csharp/src/controllers/MessagesController.cs]: Is this only protecting against the user's own code changes, or do we want to make clear that you can reach this point (conversationUpdated for 1:1 with the bot) more than once?

Comment (aosolis) [reply template-bot-master-csharp/src/controllers/MessagesController.cs]: Clarified the comment.

File: template-bot-master-csharp/src/controllers/MessagesController.cs [hunk 3]
```diff
@@ -136,12 +140,63 @@ private Activity HandleSystemMessage(Activity message)
...
+                    // Determine if the bot was added to the team/conversation
+                    var botId = message.Recipient.Id;
+                    var botWasAdded = message.MembersAdded.Any(member => member.Id == botId);
 
-                // Handle conversation state changes, like members being added and removed
-                // Use Activity.MembersAdded and Activity.MembersRemoved and Activity.Action for info
-                // Not available in all channels
+                    // Create the welcome message to send
+                    Activity welcomeMessage = message.CreateReply();
+                    welcomeMessage.Text = "Hello, I'm your new bot!";
```
Comment (jotrick) [inline template-bot-master-csharp/src/controllers/MessagesController.cs]: Reason for not using something like String.BotWelcomeMessage?

File: template-bot-master-csharp/src/controllers/MessagesController.cs [hunk 4]
```diff
@@ -136,12 +140,63 @@ private Activity HandleSystemMessage(Activity message)
...
+
+                        // If the user hasn't received a first-run message yet, then send a message to the user
+                        // introducing your bot and what it can do. Do NOT send this blindly, as your bot can receive
+                        // spurious conversationUpdate events, especially if you use proactive messaging.
+                        using (var scope = DialogModule.BeginLifetimeScope(Conversation.Container, message))
+                        {
+                            var address = Address.FromActivity(message);
+                            var botDataStore = scope.Resolve<IBotDataStore<BotData>>();
+                            var botData = await botDataStore.LoadAsync(address, BotStoreType.BotUserData, cancellationToken);
+                            if (!botData.GetProperty<bool>("IsFreSent"))
```
[the remark this answers is not in the document]
Comment (aosolis) [reply template-bot-master-csharp/src/controllers/MessagesController.cs]: Using a string for readability. It's very localized, so I'm not as worried.

**Pull request merged**